packages/opentelemetry-node/lib/environment.js

```diff
@@ -86,75 +47,42 @@ function setupEnvironment() {
 function restoreEnvironment() {
     Object.keys(envToRestore).forEach((k) => {
         process.env[k] = envToRestore[k];
+        delete envToRestore[k];
     });
 }
 
-/**
- * @typedef {Object} EdotEnv
- * @property {string[]} OTEL_NODE_RESOURCE_DETECTORS
- * @property {number} OTEL_METRIC_EXPORT_INTERVAL
- * @property {number} OTEL_METRIC_EXPORT_TIMEOUT
- * @property {boolean} ELASTIC_OTEL_METRICS_DISABLED
- */
-/**
- * @typedef {keyof EdotEnv} EdotEnvKey
- */
-/** @type {EdotEnv} */
-const edotEnv = {
-    // Missing OTEL_ vars from global spec and nodejs specific spec
-    OTEL_NODE_RESOURCE_DETECTORS: parseStringList(
-        process.env.OTEL_NODE_RESOURCE_DETECTORS,
-        ['all']
-    ),
-    // Note: Default values has been taken from the specs
-    // https://opentelemetry.io/docs/specs/otel/configuration/sdk-environment-variables/#periodic-exporting-metricreader
-    OTEL_METRIC_EXPORT_INTERVAL: parseNumber(
-        process.env.OTEL_METRIC_EXPORT_INTERVAL,
-        60000
-    ),
-    OTEL_METRIC_EXPORT_TIMEOUT: parseNumber(
-        process.env.OTEL_METRIC_EXPORT_TIMEOUT,
-        30000
-    ),
-    // ELASTIC_OTEL_ vars
-    ELASTIC_OTEL_METRICS_DISABLED: parseBoolean(
-        process.env.ELASTIC_OTEL_METRICS_DISABLED,
-        false
-    ),
-};
-
-/**
- * @typedef {import('@opentelemetry/core').ENVIRONMENT} OtelEnv
- */
-/**
- * @typedef {keyof OtelEnv} OtelEnvKey
- */
-const otelEnv = getEnv();
-
 /**
  * @template T
- * @typedef {T extends OtelEnvKey ? OtelEnv[T] : T extends EdotEnvKey ? EdotEnv[T] : never} EnvValue<T>
- */
-/**
- * @template {OtelEnvKey | EdotEnvKey} T
- * Returns the value of the env var already parsed to the proper type. If
- * the variable is not defined it will return the default value based on
- * the environmment variables spec https://opentelemetry.io/docs/specs/otel/configuration/sdk-environment-variables/
- * @param {T} name
- * @returns {EnvValue<T>}
+ * @param {(name: string) => T} getterFn
+ * @returns {(name: string, defaultValue?: T) => T}
  */
-function getEnvVar(name) {
-    if (name in otelEnv) {
-        // @ts-ignore -- T is {keyof OtelEnv} but not sure how to make TS infer that
-        return otelEnv[name];
-    }
+function makeEnvVarGetter(getterFn) {
```

Comment:
Hrm. AFAICT, the only reason for the `isStashed` handling in here is because in elastic-node-sdk.js we are currently calling `setupEnvironment` *before* the call the `resolveDetectors` that uses the `OTEL_NODE_RESOURCE_DETECTORS` envvar value.

What if we move `setupEnvironment()` to be *just before* the `super(configuration)` call. I think that was the original intent.  The only reason the `setupEnvironment / restoreEnvironment` things are done is to tweak the env for when the NodeSDK constructor runs. (Possibly also its `.start()` as well.)

Then if the `isStashed` stuff isn't needed, we could drop the `getEnv*` functions and just use the `get*FromEnv` functions from `@opentelemetry/core` directly, right?

Reply:
See commit 3c28ba4 for an attempt at this.